tests/shippers/test_logstash.py

```diff
@@ -0,0 +1,193 @@
+    assert output_type == "logstash"
+    assert event_payload == _dummy_event
+
+
+@pytest.mark.unit
+class TestLogstashShipper(TestCase):
+    @responses.activate
+    def test_send_successful(self) -> None:
+        def request_callback(request: PreparedRequest) -> tuple[int, dict[Any, Any], str]:
+            _payload = []
+            assert request.headers["Content-Encoding"] == "gzip"
+            assert request.headers["Content-Type"] == "application/x-ndjson"
+            assert request.body is not None
+            assert isinstance(request.body, bytes)
+
+            events = gzip.decompress(request.body).decode("utf-8").split("\n")
+            for event in events:
+                _payload.append(ujson.loads(event))
+
+            expected_event = _dummy_event
+            expected_event["_id"] = "_id"
+            assert _payload == [expected_event, expected_event]
+
+            return 200, {}, "okay"
+
+        def event_id_generator(event: dict[str, Any]) -> str:
+            return "_id"
+
+        url = "http://logstash_url"
+        responses.add_callback(responses.PUT, url, callback=request_callback)
+        logstash_shipper = LogstashShipper(logstash_url=url, max_batch_size=2)
+        logstash_shipper.set_event_id_generator(event_id_generator)
+        logstash_shipper.send(_dummy_event)
+        logstash_shipper.send(_dummy_event)
+
+    @responses.activate
+    def test_send_failures(self) -> None:
+        url = "http://logstash_url"
+        with self.subTest("Does not exceed max_retries"):
+            responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            responses.put(url=url, status=200)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+        with self.subTest("Exceeds max retries, replay handler set"):
+            for i in range(_MAX_RETRIES):
+                responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            replay_handler = MagicMock(side_effect=_dummy_replay_handler)
+            logstash_shipper.set_replay_handler(replay_handler)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+            replay_handler.assert_called_once_with("logstash", {}, _dummy_event)
+        with self.subTest("Exceeds max retries, replay handler not set"):
+            for i in range(_MAX_RETRIES):
+                responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            replay_handler = MagicMock(side_effect=_dummy_replay_handler)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+            replay_handler.assert_not_called()
+        with self.subTest("Authentication error, request is not retried"):
```

Comment:
that's tricky: the replay handler includes the `output_args`, so the `username` and `password`, but if they are wrong it's more likely because they were not properly set in the config file. if we send to the replay queue, the only way for those events to go through is to have in logstash the "wrong" credentials, and I see it's not what it's going to happen

but I wonder if having the replay queue empty is very specific to `401`. is there any other http code that will end up in not sending anything to the replay queue?

for me that should be rather an exceptional case. and indeed we could send also with `401`: once the user will set the replay queue as trigger of the lambda, the message will fails three times and end up in the DLQ.

not sending to the replay queue means those events are lost forever

Reply:
> that's tricky: the replay handler includes the output_args, so the username and password, but if they are wrong it's more likely because they were not properly set in the config file. if we send to the replay queue, the only way for those events to go through is to have in logstash the "wrong" credentials, and I see it's not what it's going to happen

With the current implementation, this is not correct. For Logstash, nothing is sent into `output_args`, and only `es_datastream_name` is sent for Elasticsearch outputs.
Regarding 401 errors, replayed events will go through as far as the config is fixed by the user, before replaying the events. This is true for both Elasticsearch and Logstash, from what I could see

> not sending to the replay queue means those events are lost forever
Apart from what mentioned above, I'd send failed events to the replay queue, regardless of the type of error.